telegram/ext/builders.py

```diff
@@ -0,0 +1,1338 @@
+# it under the terms of the GNU Lesser Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser Public License for more details.
+#
+# You should have received a copy of the GNU Lesser Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+#
+# Some of the type hints are just ridiculously long ...
+# flake8: noqa: E501
+# pylint: disable=C0301
+"""This module contains the Builder classes for the telegram.ext module."""
+import warnings
+from queue import Queue
+from threading import Event
+from typing import (
+    TypeVar,
+    Generic,
+    TYPE_CHECKING,
+    Callable,
+    Any,
+    Dict,
+    Union,
+    Optional,
+    overload,
+    cast,
+    Type,
+)
+
+from telegram import Bot
+from telegram.ext import Dispatcher, JobQueue, Updater, ExtBot, ContextTypes
+from telegram.ext.utils.types import CCT, UD, CD, BD, BT, DefaultContextType, JQ, PT
+from telegram.utils.request import Request
+
+if TYPE_CHECKING:
+    from telegram.ext import (
+        Defaults,
+        BasePersistence,
+        CallbackContext,
+    )
+
+# Type hinting is a bit complicated here because we try to get to a sane level of
+# leveraging generics and therefore need a number of type variables.
+ODT = TypeVar('ODT', bound=Union[None, Dispatcher])
+DT = TypeVar('DT', bound=Dispatcher)
+InBT = TypeVar('InBT', bound=Bot)
+InJQ = TypeVar('InJQ', bound=Union[None, JobQueue])
+InPT = TypeVar('InPT', bound=Union[None, 'BasePersistence'])
+InDT = TypeVar('InDT', bound=Union[None, Dispatcher])
+InCCT = TypeVar('InCCT', bound='CallbackContext')
+InUD = TypeVar('InUD')
+InCD = TypeVar('InCD')
+InBD = TypeVar('InBD')
+DefCCT = DefaultContextType  # type: ignore[misc]
+BuilderType = TypeVar('BuilderType', bound='_BaseBuilder')
+CT = TypeVar('CT', bound=Callable[..., Any])
+
+if TYPE_CHECKING:
+    InitBaseBuilder = _BaseBuilder[  # noqa: F821  # pylint: disable=E0601
+        Dispatcher[ExtBot, DefCCT, Dict, Dict, Dict, JobQueue, None],
+        ExtBot,
+        DefCCT,
+        Dict,
+        Dict,
+        Dict,
+        JobQueue,
+        None,
+    ]
+    InitUpdaterBuilder = UpdaterBuilder[  # noqa: F821  # pylint: disable=E0601
+        Dispatcher[ExtBot, DefCCT, Dict, Dict, Dict, JobQueue, None],
+        ExtBot,
+        DefCCT,
+        Dict,
+        Dict,
+        Dict,
+        JobQueue,
+        None,
+    ]
+    InitDispatcherBuilder = DispatcherBuilder[  # noqa: F821  # pylint: disable=E0601
+        Dispatcher[ExtBot, DefCCT, Dict, Dict, Dict, JobQueue, None],
+        ExtBot,
+        DefCCT,
+        Dict,
+        Dict,
+        Dict,
+        JobQueue,
+        None,
+    ]
+
+
+def check_if_already_set(func: CT) -> CT:
```

Comment:
Overriding just doesn't make any sense at all IMO, so from my pov it made sense to include this check in the flow that makes sure that the input can be reasonably be handled. But it would save same LOC …